parlai/core/torch_generator_agent.py

```diff
@@ -1584,7 +1626,12 @@ def select_paths(self, logprobs, prior_scores, current_length):
         tok_scores, tok_ids = logprobs.max(1)
         best_scores = tok_scores + prior_scores
         hyp_ids = torch.arange(logprobs.size(0)).to(logprobs.device)
-        return (hyp_ids, tok_ids, best_scores)
+
+        vocab_size = logprobs.size()[1]
+        ranks = logprobs.scatter(1, logprobs.argsort(descending=True), torch.arange(vocab_size).type(torch.FloatTensor).to(logprobs.device)[None,:].repeat((self.beam_size,1)) + 1)
+        tok_ranks = torch.gather(ranks, 1, tok_ids[None, :]).view(-1)
+
+        return (hyp_ids, tok_ids, best_scores, tok_scores.view(-1), tok_ranks)
```

Comment:
ah I would prefer if we didn't have to change the signature of select_paths. Is there a way that is possible?

Reply:
I don't think so. Since select paths is where token level probabilities are actually computed, we need some way to get them out of this method and persist them through the search process. The two approaches that come to mind are updating the type signature or have a side effect inside `select_paths` where we update a property attached to the `TreeSearch` object with the token level probabilities. If we do the latter for token probabilities, we should do it for all of these fields though, to keep these consistent. The latter option seems easier for now. There's only one callsite for `select_paths` (in `TreeSearch::advance`), so I am not clear as to why we don't want to change this signature. Why don't you want to change it?

Reply:
Only reason I'm resistant is that it adds more overhead to implementing a novel `select_paths` implementation.

Okay, this is fine. Please refactor it to return a nice object instead of an ugly 5-tuple.